src/bin/commands/demux.rs

```diff
@@ -252,25 +283,43 @@ impl Iterator for ReadSetIterator {
         if let Some(rec) = self.source.next() {
             let mut segments = Vec::with_capacity(self.read_structure.number_of_segments());
             let next_fq_rec = rec.expect("Unexpected error parsing FASTQs");
-            let read_name = next_fq_rec.head().to_vec();
-
-            for read_segment in self.read_structure.iter() {
-                let (seq, quals) = read_segment
-                    .extract_bases_and_quals(next_fq_rec.seq(), next_fq_rec.qual())
-                    .unwrap_or_else(|e| {
+            let read_name = next_fq_rec.head();
+            let bases = next_fq_rec.seq();
+            let quals = next_fq_rec.qual();
+            for (read_segment_index, read_segment) in self.read_structure.iter().enumerate() {
+                // Check that we have enough bases for this segment. For variable length segments,
+                // we must have at least one base.
+                if self.skip_reasons.contains(&SkipReason::TooFewBases)
+                    && bases.len() < read_segment.length().unwrap_or(1)
+                {
+                    return Some(ReadSet {
+                        header: read_name.to_vec(),
+                        segments: vec![],
+                        skip_reason: Some(SkipReason::TooFewBases),
+                    });
+                }
```

Comment:
A few thoughts:

1. Is this check doing the right thing?  It's just comparing the read length to the segment length ... don't you either need to compare the sum of segment lengths _or_ compare the the segment length to `bases.len() - read_segment.offset()`
2. I think it would be nicest to boost this out of the loop and do something like:

```rust
let min_len = read_structure.iter().map(|s| s.length().unwrap_or(1).sum();
if bases.len() < min_len {
    if self.skip_reasons.contains(&SkipReason::TooFewBases) {
        return Some(....)
    }
    else {
        panic!("Read {} had too few bases to demux {} vs. {} needed.", ....);
    }
}
```